engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala

```
@@ -1,84 +1,98 @@
-                                        workflowInputsAux: Map[Int, WorkflowJson],
-                                        workflowOptions: Option[WorkflowOptionsJson],
-                                        customLabels: Option[WorkflowJson],
-                                        zippedImports: Option[Array[Byte]],
-                                        warnings: Seq[String])
+final case class PartialWorkflowSources(workflowSource: Option[WorkflowSource] = None,
+                                        workflowType: Option[WorkflowType] = None,
+                                        workflowTypeVersion: Option[WorkflowTypeVersion] = None,
+                                        workflowInputs: Vector[WorkflowJson] = Vector.empty,
+                                        workflowInputsAux: Map[Int, WorkflowJson] = Map.empty,
+                                        workflowOptions: Option[WorkflowOptionsJson] = None,
+                                        customLabels: Option[WorkflowJson] = None,
+                                        zippedImports: Option[Array[Byte]] = None,
+                                        warnings: Seq[String] = List.empty)
 
 object PartialWorkflowSources {
   val log = LoggerFactory.getLogger(classOf[PartialWorkflowSources])
 
-  def empty = PartialWorkflowSources(
-    workflowSource = None,
-    workflowType = None,
-    workflowTypeVersion = None,
-    workflowInputs = Vector.empty,
-    workflowInputsAux = Map.empty,
-    workflowOptions = None,
-    customLabels = None,
-    zippedImports = None,
-    warnings = Vector.empty
-  )
-
+  implicit val partialWorkflowSourcesMonoid = new Monoid[PartialWorkflowSources] {
+    override def empty: PartialWorkflowSources = PartialWorkflowSources()
+    
+    override def combine(x: PartialWorkflowSources, y: PartialWorkflowSources): PartialWorkflowSources = {
+      x.copy(
+        workflowSource = x.workflowSource.orElse(y.workflowSource),
```

ToL: I wonder whether combining two `Some`s should be an error, here and in some of the other fields below

---

why?

---

because having two workflow sources is an error

---

> ToL: I wonder whether combining two Somes should be an error, here and in some of the other fields below

I'm guessing you're saying one might want `PartialWorkflowSources(workflowSource = "wf 1")` combined with `PartialWorkflowSources(workflowSource = "wf 2")` to be an error. Instead as is the result would be `PartialWorkflowSources(workflowSource = "wf 1")` and "wf 2" is ignored.

Is there an existing Monoid[Option]-ish function of ExclusiveOr-or-Error?

---

I agree, but I don't think this is technically possible since the form data is a `Map[String, ByteString]`, so there can't possibly be 2 "workflowSource"s, or anything else for that matter.

---

if combining two `Some`s is an error I have to wonder if this is more a big `orElse` than a monoid

---

You're right it's probably not a mathematically correct monoid but the result is still correct and can be `foldMap`ed with `ErrorOr` which makes things more concise IMO.

---

It's because "it can't happen" that I was advocating for an exception (which should never be thrown)

---

(because if we're wrong, or our assumptions about how it gets called change, we definitely want it to explode, rather than silently hiding the error)